latex/Inhalt/implementierung.tex

```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
+
+\subsection{CP/M Emulation}
+
+Um den Emulator zu testen, wurden einige frei verfügbare Programme genutzt. Diese wurden allerdings für das CP/M Betriebssystem geschrieben, welches ab 1974 unter anderem für den Intel 8080 entwickelt wurde und auf diversen Rechnern für Endanwender zum Einsatz kam.
+
+Ein wesentliches Feature, was diese Programme nutzen, sind die BDOS Syscalls, die auf einem CP/M System das Ausgeben von Text in einem Terminal ermöglichen \cite{cpmSyscalls}.
+
+\begin{minted}{asm}
+    LD DE   ;parameter
+    LD C    ;function
+    CALL 5  ;syscall routine
+\end{minted}
+
+Für den Aufruf eines Syscalls muss die Funktion, die aufgerufen wird, in das Register C geschrieben werden. Mögliche Parameter können im Register DE abgelegt werden. Der Syscall kann nun durch einen Sprung auf Adresse 5 ausgeführt werden.
+
+Relevant für die Programme, die zum Testen genutzt werden, sind lediglich die Syscalls 2 (Zeichen-Ausgabe) und 9 (String-Ausgabe). Die Emulation dieser Syscalls ist vergleichsweise simpel.
+
+\begin{minted}{typescript}
+    if (this._cpmMode) {
+      // Patch memory for BDOS syscalls
+      this._emulatorMemory[5] = 0xC9; // RET
+      this._emulator.pc = 0x100;
+    }
+\end{minted}
+
+Beim Start eines Programms, was im CP/M Modus ausgeführt wird, wird der Speicher an der Stelle 5 mit einer Return-Instruktion überschrieben. Das sorgt dafür, dass das Programm bei jedem Syscall sofort wieder an die ursprüngliche Stelle zurückspringt. Zusätzlich wird der Program Counter beim Start direkt auf 0x100 gesetzt, da CP/M davon ausgeht, dass Nutzerprogramme an dieser Stelle beginnen.
+
+Im \ts{cpuStep()} wird nun folgender Code ausgeführt:
+
+\begin{minted}{typescript}
+    if (this.emulator.pc == 0x05) {
+        // Emulate CP/M BDOS syscalls
+        const syscall = registers.c;
+        switch(syscall) {
+            case 2: {
+                // BDOS function 2 (C_WRITE) - Console output
+                const char = registers.e;
+                console.log(String.fromCharCode(char));
+                break;
+            }
+            case 9: {
+                // BDOS function 9 (C_WRITESTR) - Output string
+                let address = largeRegisters.d;
+                let currentChar = "";
+                let fullString = "";
+                while (currentChar != "$") {
+                    fullString += currentChar;
+                    const charCode = emulatorMemory[address++];
+                    currentChar = String.fromCharCode(charCode);
+                }
+                console.log(fullString);
+                break;
+            }
+        }
+    }
+\end{minted}
+
+Sobald der Program Counter den Wert 5 hat, das Programm also an die Syscall Adresse gesprungen ist, wird dieser emuliert. Beim Syscall 2 wird das Zeichen, was sich Register E befindet, ausgegeben. Syscall 9 gibt den String, der sich an der Speicheradresse aus Register DE befindet, so lange aus, bis das Zeichen \enquote{\$} auftritt.
```

Comment:
"wird dieser emuliert", what does "dieser" refer to? The code, or the syscall? Should be clarified

"was sich in Register E"